### ISSUE TITLE
Format json-pointer: uri fragment, json pointer or both

### ISSUE THREAD

**Issue by CroniD** [json-schema-org/json-schema-spec#328]
Hi there,

first: great work with draft-06. I really like it and I'm about to update my schemes for it. Looks really good so far.

But one thing is unclear to me:

> 8.3.9.  json-pointer
> 
>    This attribute applies to string instances.
> 
>    A string instance is valid against this attribute if it is a valid
>    JSON Pointer, according to [RFC6901]

This means the "json pointer" representation or the uri fragment one, or both? Some validators like ajv (@epoberezkin) allow both versions, but the term "valid JSON pointer" may suggest otherwise. I guess most people will expect both representations to work, but as it is written above it may not be intended (strictly speaking)?
Implementationwise it should be a difference, because an uri fragment may contain percent-encoded characters, while the other one doesn't.

Best regards

**Comment by handrews** [comment]
My intention with that format was an actual JSON Pointer ("/some/property"), and not a fragment that happens to be a JSON Pointer ("#/some/property").  The fragment would be format uri-reference.

Since RFC 6901 describes the fragment encoding as well, I can see how this would be confusing.

**Comment by epoberezkin** [comment]
@handrews `uri-reference` is wider than fragment that is JSON pointer. Maybe there should be two formats for JSON pointer instead of one...

**Comment by handrews** [comment]
@epoberezkin I don't see any need for formats for specific fragment-only URI reference syntaxes.  Or any other subset of `uri-reference`.  If you really want to ensure that a URI reference is a fragment, then use:

```JSON
{
    "type": "string",
    "format": "uri-reference",
    "pattern": "^#"
}
```

Since fragments are interpreted based on the media type of the resourced being referenced, whether or not that fragment can be (or must be) a JSON Pointer is determined by the media type.  In the case where a media type (such as `application/schema+json`) supports multiple fragment resolutions, then you can use a more specific pattern to ensure only JSON Pointer.  For instance, the following is sufficient for fragments referencing JSON schemas:

```JSON
{
    "type": "string",
    "format": "uri-reference",
    "pattern": "^#/"
}
```

(note the extra forward slash in the pattern)

**Comment by CroniD** [comment]
Thx for your replies.

Do you think it's worth to put a little note to 8.3.9. stating that the fragment representation is excluded?

OR

RFC6901 doesn't use the term "valid Json Pointer". Maybe use "JSON String representation" instead (be close to RFC6901 as possible)?

Are new tests needed to check that? As I can see there are some tests for the json-pointer format.

Best regards

**Comment by handrews** [comment]
@CroniD yes I think some clarification is in order, leave this issue open and either submit a PR or I'll get around to it eventually (I'm catching up from vacation and also have a cold so I'm not doing much this week).

I'm sure @Julian would welcome PRs on the test suite if there is a gap.

**Comment by Julian** [comment]
He would!

**Comment by CroniD** [comment]
Hi all,

tests are written.

@handrews I'm not very good at writing a specification, therefore I'll leave this matter to another one. Best wishes, I hope you will get better soon.


Thank you all. :)

**Comment by handrews** [comment]
Further evidence that this requires clarification in the standard (and probably more on the web site, but at least a line about it in the spec seems necessary).

https://groups.google.com/d/msg/json-schema/gP6T8dEVJFQ/Ek3yhbAlBAAJ

**Comment by handrews** [comment]
Fixed in #354.
